Cargo.toml

```diff
@@ -26,6 +26,7 @@ regex = "1"
 base64 = "0.13"
 lazy_static = "1"
 bimap = { version = "0.6.3", features = ["serde"] }
+take_mut = "0.2.2"
```

Comment:
Can we avoid taking a dependency on this crate? I think it's only used in the `merge_field` function which can be rewritten as follows without this crate:

```rust
pub fn merge_field<'a, T>(
    target_fields: &mut Vec<Field<'a, T>>,
    mut field: Field<'a, T>,
    type_name: &str,
    field_map: &HashMap<String, __Field>,
) -> Result<(), String>
where
    T: Text<'a> + Eq + AsRef<str> + std::fmt::Debug + Clone,
{
    let mut matching_field_index = target_fields.len();
    for (i, target) in target_fields.iter().enumerate() {
        if alias_or_name(target) == alias_or_name(&field) {
            matching_field_index = i;
            break;
        }
    }

    if matching_field_index == target_fields.len() {
        target_fields.push(field);
        return Ok(());
    }

    let matching_field = &target_fields[matching_field_index];

    can_fields_merge(matching_field, &field, type_name, field_map)?;

    field.position = field.position.min(matching_field.position);

    field.selection_set.span =
        min_encapsulating_span(field.selection_set.span, matching_field.selection_set.span);
    field
        .selection_set
        .items
        .extend(matching_field.selection_set.items.clone());
    target_fields[matching_field_index] = field;

    Ok(())
}
```

Reply:
We could, but we'd have to clone the selection set items for the matching field.

I figured this dependency would let us avoid having to clone any extra data, as it simultaneously consumes the matching field and replaces it without extra allocations.

Reply:
Alternatively, if we want a zero-dependency solution, we could either:

1. use `unsafe` to consume `matching_field` from the vector, then make the guarantee that we're placing a new value back in the original position.

2. use `swap_remove` to consume the value, create the new value and push into the vector, then `swap` to move the new value back to the original position.

Reply:
We could also just copy the `take_mut` implementation since it has zero dependencies.

```rs
pub fn take<T, F>(mut_ref: &mut T, closure: F)
  where F: FnOnce(T) -> T {
    use std::ptr;

    unsafe {
        let old_t = ptr::read(mut_ref);
        let new_t = panic::catch_unwind(panic::AssertUnwindSafe(|| closure(old_t)))
            .unwrap_or_else(|_| ::std::process::abort());
        ptr::write(mut_ref, new_t);
    }
}
```

Reply:
Yeah, both these are valid approaches. I'm inclined towards using `swap_remove`, although haven't looked at whether we'd be able to create a temporary dummy `Field`.

Reply:
Lets avoid `unsafe` if possible.

Reply:
The general idea I went with was to use the incoming `field`'s memory as the pre-allocated `Field`, then merge the matching field onto `field`.

Either way, we have to simultaneously consume and produce the new field value, which isn't possible with a standard replacement.

Reply:
The swapping approach would look like:

```rs
    let Some(matching_idx) = target_fields
        .iter_mut()
        .position(|target| alias_or_name(target) == alias_or_name(&field))
    else {
        target_fields.push(field);
        return Ok(());
    };

    can_fields_merge(&target_fields[matching_idx], &field, type_name, field_map)?;
    
    let matching_field = target_fields.swap_remove(matching_idx);

    field.position = field.position.min(matching_field.position);

    field.selection_set.span =
        min_encapsulating_span(field.selection_set.span, matching_field.selection_set.span);

    // Subfields will be normalized and properly merged on a later pass.
    field
        .selection_set
        .items
        .extend(matching_field.selection_set.items);
    
    target_fields.push(field);
    let last_idx = target_fields.len() - 1;
    target_fields.swap(last_idx, matching_idx);
```

But I worry it's a little more confusing as to why we're swapping elements around, plus the small cost of moving elements around.

Reply:
> Lets avoid `unsafe` if possible.

That's why I used `take_mut`, since it's extremely lightweight, has no dependencies, well tested to cover the required `unsafe` usage, and does exactly what we want declaratively.

Reply:
The original reason I was trying to avoid `take_mut` was to keep dependencies to a minimum but I just found out that it [aborts the process](https://github.com/Sgeo/take_mut/blob/1bd70d842c6febcd16ec1fe3a954a84032b89f52/src/lib.rs#L38) if the closure given to `take` method panics. E.g. the following will crash the Postgres process:

```rust
    let mut i = 0;
    take_mut::take(&mut i, |ii| {
        panic!("panic from inside take");
    });
```

`pgrx`'s machinery does handle panics safely e.g. this doesn't crash the Postgres process:

```rust
panic!("just a panic");
```

but `pgrx` can't do anything about a call to `::std::process::abort` by an extension. Honestly, I'd rather just `clone` and pay a bit of a performance penalty than introduce another dependency.

Reply:
Ah you're right, that's not desired behaviour. I've updated the PR to reflect the decision!